controller/tests/models/domain/imagefilter_test.py

```diff
@@ -0,0 +1,191 @@
+"""Tests of Docker image tag parsing and analysis."""
+
+from __future__ import annotations
+
+from datetime import UTC, datetime, timedelta
+from random import SystemRandom
+
+from semver import Version
+
+from controller.models.domain.imagefilterpolicy import (
+    ImageFilterPolicy,
+    RSPImageFilterPolicy,
+)
+from controller.models.domain.rsptag import (
+    RSPImageTagCollection,
+    RSPImageType,
+)
+
+
+def test_imagefilter() -> None:
+    """Test behavior of an RSPImageTagCollection object under a filter."""
+
+    # Make some weekly and daily tags.  Because we can run this test at
+    # an arbitrary time...we need to dynamically generate these.
+    def to_weekly_tag(dt: datetime) -> str:
+        ic = dt.isocalendar()
+        icy = str(ic.year).zfill(2)
+        icw = str(ic.week).zfill(2)
+        return f"w_{icy}_{icw}"
+
+    def to_daily_tag(dt: datetime) -> str:
+        mo = str(dt.month).zfill(2)
+        dy = str(dt.day).zfill(2)
+        return f"d_{dt.year}_{mo}_{dy}"
+
+    # Because image ages are calculated based on the tags, and are
+    # therefore calculated as if they were born at midnight UTC, you
+    # might have one fewer image than you expect in a category depending
+    # on local time if you're west of UTC, or one more if you're east of UTC.
```

Comment:
If that's the case there's some serious flaw in the code somewhere because nothing in either the controller or in the tests should be based on local time zones. Every operation should be done in UTC from beginning to end, so nothing related to time zones should matter.

There therefore shouldn't be any ranges; we should be able to determine the expected output exactly.

Reply:
I think the problem is that now() might be after midnight the next day, or before midnight of the day the image was presumably minted, depending on where you actually are.  That means an image can be a negative duration old, and "yesterday" and "today" don't mean (with respect to the image age) what they actually should.

I mean, I think maybe where I'm going with this is that maybe "age", since we can't really get the image age, the best we can do is "midnight UTC of the date in the tag" or perhaps "upload date" or "distrib tag creation date", is not a good filter choice.  Maybe we really can do everything we need to do with image count and semver.

Reply:
Oh, nm, now I get why I shouldn't default it in `filter()`